Mathlib/Algebra/Order/BigOperators/Expect.lean

```diff
@@ -0,0 +1,186 @@
+
+lemma abs_expect_le_expect_abs (s : Finset ι) (f : ι → α) : |𝔼 i ∈ s, f i| ≤ 𝔼 i ∈ s, |f i| :=
+  le_expect_of_subadditive _ abs_zero abs_add (fun _ _ ↦ abs_nnqsmul _ _) _ _
+
+end LinearOrderedAddCommGroup
+
+section LinearOrderedCommSemiring
+variable [LinearOrderedCommSemiring R] [ExistsAddOfLE R] [Module ℚ≥0 R] [PosSMulMono ℚ≥0 R]
+
+/-- **Cauchy-Schwarz inequality** in terms of `Finset.expect`. -/
+lemma expect_mul_sq_le_sq_mul_sq (s : Finset ι) (f g : ι → R) :
+    (𝔼 i ∈ s, f i * g i) ^ 2 ≤ (𝔼 i ∈ s, f i ^ 2) * 𝔼 i ∈ s, g i ^ 2 := by
+  simp only [expect, smul_pow, inv_pow, smul_mul_smul_comm, ← sq]
+  gcongr
+  exact sum_mul_sq_le_sq_mul_sq ..
+
+end LinearOrderedCommSemiring
+end Finset
+
+open Finset
+
+namespace Fintype
+variable [Fintype ι] [Fintype κ]
+
+section OrderedAddCommMonoid
+variable [OrderedAddCommMonoid α] [Module ℚ≥0 α] {f : ι → α}
+
+lemma expect_eq_zero_iff_of_nonneg [Nonempty ι] (hf : 0 ≤ f) : 𝔼 i, f i = 0 ↔ f = 0 := by
+  simp [expect, sum_eq_zero_iff_of_nonneg hf, univ_nonempty.ne_empty]
+
+lemma expect_eq_zero_iff_of_nonpos [Nonempty ι] (hf : f ≤ 0) : 𝔼 i, f i = 0 ↔ f = 0 := by
+  simp [expect, sum_eq_zero_iff_of_nonpos hf, univ_nonempty.ne_empty]
+
+end OrderedAddCommMonoid
+end Fintype
+
+open Finset
+
+namespace Mathlib.Meta.Positivity
+open Qq Lean Meta Finset
+open scoped BigOperators
+
+/-- Positivity extension for `Finset.expect`. -/
+@[positivity Finset.expect _ _]
+def evalFinsetExpect : PositivityExt where eval {u α} zα pα e := do
+  match e with
+  | ~q(@Finset.expect $ι _ $instα $instmod $s $f) =>
+    let (lhs, _, (rhs : Q($α))) ← lambdaMetaTelescope f
+    let so : Option Q(Finset.Nonempty $s) ← do
+      try
+        let _fi ← synthInstanceQ q(Fintype $ι)
+        let _no ← synthInstanceQ q(Nonempty $ι)
+        match s with
+        | ~q(@univ _ $fi) => pure (some q(Finset.univ_nonempty (α := $ι)))
+        | _ => pure none
+      catch _ => do
+        let .some fv ← findLocalDeclWithType? q(Finset.Nonempty $s) | pure none
+        pure (some (.fvar fv))
+    match ← core zα pα rhs, so with
+    | .nonnegative pb, _ => do
+      let instαordmon ← synthInstanceQ q(OrderedAddCommMonoid $α)
+      let instαordsmul ← synthInstanceQ q(PosSMulMono ℚ≥0 $α)
+      assumeInstancesCommute
+      let pr : Q(∀ i, 0 ≤ $f i) ← mkLambdaFVars lhs pb
+      return .nonnegative q(@expect_nonneg $ι $α $instαordmon $instmod $s $f $instαordsmul
+        fun i _ ↦ $pr i)
+    | .positive pb, .some (fi : Q(Finset.Nonempty $s)) => do
+      let instαordmon ← synthInstanceQ q(OrderedCancelAddCommMonoid $α)
+      let instαordsmul ← synthInstanceQ q(PosSMulStrictMono ℚ≥0 $α)
+      assumeInstancesCommute
+      let pr : Q(∀ i, 0 < $f i) ← mkLambdaFVars lhs pb
+      return .positive q(@expect_pos $ι $α $instαordmon $instmod $s $f $instαordsmul
+        (fun i _ ↦ $pr i) $fi)
+    | _, _ => pure .none
+  | _ => throwError "not Finset.expect"
+
+example (n : ℕ) (a : ℕ → ℚ) : 0 ≤ 𝔼 j ∈ range n, a j^2 := by positivity
```

Comment:
Could be useful to have a separate file for testing positivity extensions...

Reply:
The point of having the tests right after the extensions is that the way positivity works implies that it works cumulatively: If positivity proves a goal, then importing more things will not make it not prove that goal anymore

Reply:
This sounds like it's an argument for centralising the extensions! The idea is that if they're all in one place, it's easy to test all the extensions simultaneously - we already have a test file for positivity, and I'm saying we should extend/imitate that one, rather than placing them sporadically throughout mathlib

Reply:
> This sounds like it's an argument for centralising the extensions!

Can you expand? I don't understand how "positivity extensions work cumulatively" implies "positivity extensions should all be defined in the same place"

Reply:
I'm not saying they should all be defined in the same place! I'm saying they should all be _tested_ in the same place, and cumulativity implies this is just as fair a way of testing. The locations they're currently defined is better
(I realise now that the first sentence in my last message made this unclear - sorry!)

Reply:
> cumulativity implies this is just as fair a way of testing

Maybe, but testing in the individual files means you get much faster feedback when debugging the extension. This is nonnegligible

Reply:
Note that what you're suggesting is inconsistent with the way positivity extensions are already tested, inconsistent with the rest of mathlib, and inconsistent with general software design principles. What is so special about the way that this particular positivity extension works that means it deserves to be treated differently?

Reply:
> inconsistent with the way positivity extensions are already tested

Factually incorrect: Grepping `example.*positivity` returns examples in 12 files that are not `test.positivity`. That's roughly a third of the files containing positivity extensions

> inconsistent with the rest of mathlib

Explained by the fact that positivity extensions are cumulative, which is a pretty rare feature among mathlib tactics

> inconsistent with general software design principles

Citation needed, but sure I'll concede

Reply:
Really what I'm saying is: two thirds of the positivity extensions are tested in one place (the `test/` folder), and these ones should be tested there too.

Reply:
Anyway I think it's bad to have tests scattered like this, but I won't block the PR on it

Reply:
I think it makes a lot of sense scattered, so 🤷 Let's find something more interesting to talk about